qiskit/primitives/sampler.py

```diff
@@ -158,6 +158,12 @@ def _run(
     def _preprocess_circuit(circuit: QuantumCircuit):
         circuit = init_circuit(circuit)
         q_c_mapping = final_measurement_mapping(circuit)
+        if set(range(circuit.num_clbits)) != set(q_c_mapping.values()):
```

Comment:
This looks like it is still going to raise if the number of clbits is not equal to the number of final measurements.  That is to say that if I wrote to all my bits via mid-circuit measurements then this would be triggered.

Reply:
This is part of the reference implementation based on `quantum_info.Statevector`. It requires final measurements.

Reply:
I updated the title of this PR to mention `BaseSampler`

Reply:
But then you are blocking real-world use cases for a simulation method that does not scale.  Does not seem like a fruitful direction to me.

Reply:
The reference implementation was designed to be as simple as possible. Runtime primitives and Aer primitives are more appropriate for scaling, e.g., https://github.com/Qiskit/qiskit-aer/pull/1531
This is similar to the relationship of `BasicAer` and `Aer` simulator. The reference implementation of primitives corresponds to `BasicAer` and Aer primitives corresponds to `Aer` simulator.

Reply:
That is another way to think about it; Just overload all in Runtime.  In some sense that is what would probably need to be done because a state vector simulator is quite far removed from a real hardware sampling routine

Reply:
There is a PR #8668 working on backend-based Sampler. It is closer to real hardware sampling. Once it's merged, we can discuss whether we need the reference implementation.

Reply:
This check is only part of the reference implementation of `Sampler` and not part of `BaseSampler`. So, other Sampler implementations does use this check.

Reply:
The reference implementation `Sampler` needs this check. So, I have to put it in `BaseSampler` (abstract class) or `Sampler` (reference implementation). I moved it from `Sampler` to `BaseSampler` in #8668. I want to revert it in this PR. If this PR is not merged, the check stays in `BaseSampler` and raises an error for all implementations not only the reference impl but also `qiskit-ibm-runtime`.